rust/src/librarydb.rs

```diff
@@ -0,0 +1,504 @@
+        };
+        if let Some(position) = checked_position {
+            if position >= entry.data_start {
+                self.position = position;
+                return Ok(position - entry.data_start);
+            }
+        }
+        // must never become negative or overflow
+        Err(io::ErrorKind::InvalidInput.into())
+    }
+}
+
+/// LibraryFile reads the data of a library entry.
+impl io::Read for LibraryFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut library = self.arc_library.write().unwrap();
+        let end = library.entries[self.index].data_end;
+        if self.position >= end {
+            return Ok(0);
+        }
+        let available = end - self.position;
+        library.library_file.seek(SeekFrom::Start(self.position))?;
+        let read_result = if available < buf.len() as u64 {
+            library
+                .library_file
+                .read(buf.split_at_mut(available as usize).0)
+        } else {
+            library.library_file.read(buf)
+        };
+        if let Ok(bytes) = read_result {
+            self.position += bytes as u64;
+        }
+        read_result
+    }
+}
+
+/// Performs a checked unsigned signed addition.
+fn checked_add_u64_i64(a: u64, b: i64) -> Option<u64> {
+    if b > 0 {
+        a.checked_add(b as u64)
+    } else if b < 0 {
+        a.checked_sub(-b as u64)
+    } else {
+        Some(a)
+    }
+}
+
+/// Gets the case insensitive representation of a path in the library database.
+fn case_insensitive_path(path: &str) -> String {
+    path.to_ascii_uppercase().replace('\\', "/")
+}
+
+/// Finds a filesystem file.
+/// dir_path is an absolute path or a path relative to the current directory.
+/// file_name is a path relative to dir_path, the normal components are searched case-insensitive (perfect match takes precedence).
+fn find_file(dir_path: &Path, file_name: &Path) -> io::Result<PathBuf> {
+    let mut path = dir_path.to_owned();
+    let components: Vec<_> = file_name.components().collect();
+    for (i, component) in components.iter().enumerate() {
+        match component {
+            Component::Normal(os_str) => {
+                if let Some(want) = os_str.to_str() {
+                    let found: Vec<_> = path
+                        .read_dir()?
+                        .filter_map(|x| x.ok()) // DirEntry
+                        .filter_map(|x| match x.file_type() {
+                            Ok(file_type) => Some((x.file_name(), file_type)),
+                            _ => None,
+                        }) // (OsString,FileType)
+                        .collect();
+                    if !found.is_empty() {
+                        if found.iter().any(|x| &x.0 == os_str) {
+                            path.push(os_str); // perfect match
+                            continue;
+                        }
+                        let want_ascii_uppercase = want.to_ascii_uppercase();
```

Comment:
Same as above, but paths are definitely not only ascii, so `to_uppercase` should be used here...

Reply:
Can you give me an example of an existing non-ascii path?
I only found ascii in the original game data.

Reply:
Paths can be absolute here, so things like ~gyerk/trúgȧmês/ja2/...

Reply:
Yes. E.g. https://github.com/ja2-stracciatella/ja2-stracciatella/issues/719 is a bug that happens because we dont handle non-ascii paths correctly...

Reply:
The `data_dir` part is not treated as case insensitive so it's not a problem.
Only the `file_name` part is treated as case insensitive.

Hmm, currently it is possible to exit `data_dir` while traversing `file_name` and that is not good.
Will make it panic in that case.